jena-db/jena-tdb2/src/main/java/org/apache/jena/tdb2/sys/DatabaseOps.java

```diff
@@ -258,46 +252,74 @@ private static void compact(DatasetGraphSwitchable container, Location loc1, Loc
+        container.execReadOnlyDatabase(()->{
+            // No active writers or promote transactions on the current database.
+            // These are held up on a lock in the switchable container.
 
-        txnMgr2.execExclusive(()->{
-            List<TransactionalComponent> externals = txnMgr.listExternals();
-            if ( ! externals.isEmpty() ) {
-                txnMgr2.modifyConfigDirect(()->externals.forEach(txnMgr2::addExternal));
-            }
+            // -- Copy the current state to the new area.
+            copyConfigFiles(loc1, loc2);
+            DatasetGraphTDB dsgCompact = StoreConnection.connectCreate(loc2).getDatasetGraphTDB();
+            CopyDSG.copy(dsgBase, dsgCompact);
 
-            // -- Switch.
-            // No transactions on either database.
-            if ( ! container.change(dsgCurrent, dsgCompact) ) {
-                Log.warn(DatabaseOps.class, "Inconistent: old datasetgraph not as expected");
-                container.set(dsgCompact);
+            if ( false ) {
+                // DEVELOMENT. FAke a long copy time in state copy.
+                System.err.println("-- Inside compact 1");
+                Lib.sleep(3_000);
+                System.err.println("-- Inside compact 2");
             }
 
-            // This switches off the source database. Exclusive mode is not exited.
-            txnMgr.startExclusiveMode();
+            TransactionCoordinator txnMgr2 = dsgCompact.getTxnSystem().getTxnMgr();
+            // Update TransactionCoordinator and switch over.
+            txnMgr2.execExclusive(()->{
+                // No active transactions in either database.
+                txnMgr2.takeOverFrom(txnMgr1);
+
+                // Copy over external transaction components.
+                txnMgr2.modifyConfigDirect(()-> {
+                    txnMgr1.listExternals().forEach(txnMgr2::addExternal);
+                    // External listeners?
+                    // (the NodeTableCache listener is not external)
+                });
+
+                // No transactions on new database 2 (not exposed yet).
+                // No writers or promote transactions on database 1.
+                // Maybe old readers on database 1.
+                // -- Switch.
+                if ( ! container.change(dsgCurrent, dsgCompact) ) {
+                    Log.warn(DatabaseOps.class, "Inconistent: old datasetgraph not as expected");
+                    container.set(dsgCompact);
+                }
+                // The compacted database is now active
+            });
+            // New database running.
+            // New transactions go to this database.
+            // Old readers continue on db1.
+
         });
 
-        // New database running.
 
+        // This switches off the source database.
+        // It waits until all transactions (readers) have finished.
+        // This call is not undone.
+        // Database1 is no longer in use.
```

Comment:
I think at this point we have two config files, and two databases. Are these left as-is, or is there a process to clean them up later?

Reply:
There's an option to delete now in the outer `compact(DatasetGraphSwitchable container, boolean shouldDeleteOld)` and Fuseki exposes this with `?deleteOld=true`. In Fuseki compaction is async tot he request - it's a background task, because it may take a while is held up by waiting for the database to be writer-free.

Another choice is to archive the old database and use as a snapshot/backup. It's fully functional and could be restored, or run elsewhere.

Reply:
Thanks!

>it's a background task, because it may take a while is held up by waiting for the database to be writer-free.

Ah, I think I saw some comments in the mailing list about compaction running in the background. Do you know if the backup appears in the Fuseki UI, under Tasks?

Reply:
>Do you know if the backup appears in the Fuseki UI, under Tasks?

Ah, it does! Great!

![image](https://user-images.githubusercontent.com/304786/180914877-246a4aa2-f03e-43ba-8fe6-e62d889d03f7.png)